pennylane_lightning/core/src/measurements/MeasurementsBase.hpp

```diff
@@ -228,6 +228,166 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
+        PL_ABORT_IF(
+            obs.getObsName().find("Hamiltonian") != std::string::npos,
+            "Hamiltonian and Sparse Hamiltonian do not support samples().");
+        std::vector<size_t> obs_wires;
+        std::vector<size_t> identity_wires;
+        auto sv = _preprocess_state(obs, obs_wires, identity_wires);
+        Derived measure(sv);
+        return measure.probs(obs_wires);
+    }
+
+    /**
+     * @brief Return samples drawn from eigenvalues of the observable
+     *
+     * @param obs The observable object to sample
+     * @param num_shots Number of shots used to generate samples
+     *
+     * @return samples of eigenvalues of the observable
+     */
+    auto sample(const Observable<StateVectorT> &obs, const size_t &num_shots)
+        -> std::vector<PrecisionT> {
+        PL_ABORT_IF(
+            obs.getObsName().find("Hamiltonian") != std::string::npos,
+            "Hamiltonian and Sparse Hamiltonian do not support samples().");
+        std::vector<size_t> obs_wires;
+        std::vector<size_t> identity_wires;
+        std::vector<size_t> shot_range = {};
+        size_t term_idx = 0;
+
+        return measure_with_samples(obs, num_shots, shot_range, term_idx);
+    }
+
+    /**
+     * @brief Return the raw basis state samples
+     *
+     * @param num_shots Number of shots used to generate samples
+     *
+     * @return the raw basis state samples
+     */
+    auto sample(const size_t &num_shots) -> std::vector<size_t> {
+        Derived measure(_statevector);
+        return measure.generate_samples(num_shots);
+    }
+
+    /**
+     * @brief Groups the eigen values of samples into a dictionary showing
+     * number of occurences for each possible outcome with the number of shots.
+     *
+     * @param obs The observable to sample
+     * @param num_shots number of wires the sampled observable was performed on
+     *
+     * @return std::unordered_map<PrecisionT, size_t> with format
+     * ``{'EigenValue': num_occurences}``
+     */
+    auto counts(const Observable<StateVectorT> &obs, const size_t &num_shots)
+        -> std::unordered_map<PrecisionT, size_t> {
+        std::unordered_map<PrecisionT, size_t> outcome_map;
+        std::vector<size_t> shot_range = {};
+        auto sample_data = sample(obs, num_shots);
+        for (size_t i = 0; i < num_shots; i++) {
+            auto key = sample_data[i];
+            auto it = outcome_map.find(key);
+            if (it != outcome_map.end()) {
+                it->second += 1;
+            } else {
+                outcome_map[key] = 1;
+            }
+        }
+        return outcome_map;
+    }
+
+    /**
+     * @brief Groups the samples into a dictionary showing number of occurences
+     * for each possible outcome with the number of shots.
+     *
+     * @param num_shots number of wires the sampled observable was performed on
+     *
+     * @return std::unordered_map<std::string, size_t> with format ``{'outcome':
+     * num_occurences}``
```

Comment:
Actually, I implemented based on the Pennylane docs. Thanks @dime10 for the information! Just Updated current implementation by using `size_t` instead of `string` here to make sure that it works fine with Catalyst. I think we can even store it into a std::vector if the key is `size_t`.

Reply:
Thanks @multiphaseCFD !